llvm/lib/CodeGen/MachineSSAUpdater.cpp

```diff
@@ -236,6 +236,19 @@ void MachineSSAUpdater::RewriteUse(MachineOperand &U) {
     NewVR = GetValueInMiddleOfBlock(UseMI->getParent());
   }
 
+  // Insert a COPY if register class contraint isn't fulfilled.
+  if (const TargetRegisterClass *RC =
+          dyn_cast_or_null<const TargetRegisterClass *>(RegAttrs.RCOrRB)) {
+    if (NewVR && !RC->hasSubClassEq(MRI->getRegClass(NewVR))) {
+      MachineBasicBlock *UseBB = UseMI->getParent();
+      MachineInstr *InsertedCopy =
+          InsertNewDef(TargetOpcode::COPY, UseBB, UseBB->getFirstNonPHI(),
```

Comment:
I'm not sure what MachineSSAUpdater requires, but it might be possible to try constrainRegClass before creating a copy (and presumably that would require changing the SSAUpdater'rs RegAttrs which may be undesirable)

Reply:
I've changed this to first try to constrain the class of NewVR using constrainRegClass. So we only add a COPY if it couldn't be constrained.

I think it is a bit hard to reason about if constraining the class is better then adding a COPY here. It probably depends on the live range and other uses of NewVR, and if passes between early-tail-duplication and the regular register coalescing would suffer from the COPY. But I also don't think it is the responsibility of tail duplication to make a deep analysis of that (or else we would need to duplicate a lot of the logic in the register coalescer/allocator here).

I also intentianally just use the original RC for the MO we are rewriting as input to the constrainRegClass. Another option would be to find the register class defined for the actual machine instruction. But then we would need to consider potential subregister uses as well. I don't think it is worth the complexity. (One could perhaps use recomputeRegClass in some more places, but afaik that is done later by the RegisterCoalescer anyway.)